extmod/modujson.c

```diff
@@ -31,25 +31,117 @@
 #include "py/parsenum.h"
 #include "py/runtime.h"
 #include "py/stream.h"
+#include "py/nlr.h"
 
 #if MICROPY_PY_UJSON
 
-STATIC mp_obj_t mod_ujson_dump(mp_obj_t obj, mp_obj_t stream) {
-    mp_get_stream_raise(stream, MP_STREAM_OP_WRITE);
-    mp_print_t print = {MP_OBJ_TO_PTR(stream), mp_stream_write_adaptor};
-    mp_obj_print_helper(&print, obj, PRINT_JSON);
+static void mod_ujson_separators(mp_obj_t separators_in, const char **item_separator, const char **key_separator) {
+    if (separators_in == mp_const_none) {
+        *item_separator = ", ";
+        *key_separator = ": ";
+    } else {
+        mp_obj_t *items;
+        size_t len;
+        mp_obj_tuple_get(separators_in, &len, &items);
+
+        if (len != 2) {
+            mp_raise_ValueError(MP_ERROR_TEXT("too many values to unpack (expected 2)"));
+        }
+
+        *item_separator = mp_obj_str_get_str(items[0]);
+        *key_separator = mp_obj_str_get_str(items[1]);
+    }
+}
+
+const char *ujson_item_separator = ", ";
+const char *ujson_key_separator = ": ";
+
+STATIC mp_obj_t mod_ujson_dump(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum {ARG_indent, ARG_separators};
+    const mp_arg_t allowed_args[] = {
+        { MP_QSTR_indent, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_rom_obj = mp_const_none} },
+        { MP_QSTR_separators, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_rom_obj = mp_const_none} },
+    };
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 2, pos_args + 2, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    if (args[ARG_indent].u_obj != mp_const_none) {
+        mp_raise_NotImplementedError(MP_ERROR_TEXT("indent is not None"));
+    }
+
+    mp_get_stream_raise(pos_args[1], MP_STREAM_OP_WRITE);
+
+    const char *old_item_separator = ujson_item_separator;
+    const char *old_key_separator = ujson_key_separator;
+    bool raise = false;
+    mod_ujson_separators(args[ARG_separators].u_obj, &ujson_item_separator, &ujson_key_separator);
+
+    nlr_buf_t nlr;
```

Comment:
what is the nlr stuff necessary for?

Reply:
The nlr stuff is necessary to reset the variables to they're
original state in case `mp_obj_print_helper` raises an exception
inside an interrupt.  

This applies to all ISR's but again could be removed since 
inside ISR's json.dump/s() should not be called.
I just did not want to leave the variables in an uncertain state.